pkg/model/keycloak_deployment.go

```diff
@@ -11,7 +12,104 @@ import (
 	"sigs.k8s.io/controller-runtime/pkg/client"
 )
 
-func KeycloakDeployment(cr *v1alpha1.Keycloak) *v13.StatefulSet {
+func getKeycloakEnv(cr *v1alpha1.Keycloak, dbSecret *v1.Secret) []v1.EnvVar {
+	env := []v1.EnvVar{
+		// Database settings
+		{
+			Name:  "DB_VENDOR",
+			Value: "POSTGRES",
+		},
+		{
+			Name:  "DB_SCHEMA",
+			Value: "public",
+		},
+		{
+			Name:  "DB_ADDR",
+			Value: PostgresqlServiceName + "." + cr.Namespace + ".svc.cluster.local",
+		},
+		{
+			Name:  "DB_DATABASE",
+			Value: GetExternalDatabaseName(dbSecret),
+		},
+		{
+			Name: "DB_USER",
+			ValueFrom: &v1.EnvVarSource{
+				SecretKeyRef: &v1.SecretKeySelector{
+					LocalObjectReference: v1.LocalObjectReference{
+						Name: DatabaseSecretName,
+					},
+					Key: DatabaseSecretUsernameProperty,
+				},
+			},
+		},
+		{
+			Name: "DB_PASSWORD",
+			ValueFrom: &v1.EnvVarSource{
+				SecretKeyRef: &v1.SecretKeySelector{
+					LocalObjectReference: v1.LocalObjectReference{
+						Name: DatabaseSecretName,
+					},
+					Key: DatabaseSecretPasswordProperty,
+				},
+			},
+		},
+		// Discovery settings
+		{
+			Name:  "NAMESPACE",
+			Value: cr.Namespace,
+		},
+		{
+			Name:  "JGROUPS_DISCOVERY_PROTOCOL",
+			Value: "dns.DNS_PING",
+		},
+		{
+			Name:  "JGROUPS_DISCOVERY_PROPERTIES",
+			Value: "dns_query=" + KeycloakDiscoveryServiceName + "." + cr.Namespace + ".svc.cluster.local",
+		},
+		// Cache settings
+		{
+			Name:  "CACHE_OWNERS_COUNT",
+			Value: "2",
+		},
+		{
+			Name:  "CACHE_OWNERS_AUTH_SESSIONS_COUNT",
+			Value: "2",
+		},
+		{
+			Name: "KEYCLOAK_USER",
+			ValueFrom: &v1.EnvVarSource{
+				SecretKeyRef: &v1.SecretKeySelector{
+					LocalObjectReference: v1.LocalObjectReference{
+						Name: "credential-" + cr.Name,
+					},
+					Key: AdminUsernameProperty,
+				},
+			},
+		},
+		{
+			Name: "KEYCLOAK_PASSWORD",
+			ValueFrom: &v1.EnvVarSource{
+				SecretKeyRef: &v1.SecretKeySelector{
+					LocalObjectReference: v1.LocalObjectReference{
+						Name: "credential-" + cr.Name,
+					},
+					Key: AdminPasswordProperty,
+				},
+			},
+		},
+		{
+			Name:  "KEYCLOAK_POSTGRESQL_SERVICE_HOST",
+			Value: PostgresqlServiceName + "." + cr.Namespace + ".svc.cluster.local",
+		},
+		{
+			Name:  "KEYCLOAK_POSTGRESQL_SERVICE_PORT",
+			Value: fmt.Sprintf("%v", GetExternalDatabasePort(dbSecret)),
+		},
```

Comment:
@pb82 Would be good to only add these when the externalDatabase option is true. Would also be good to build the name dynamically from the constant name.